### PR TITLE
In-Memory: sync RegionLabelRules from PD

### LINKED ISSUE #16606: Pull down RegionLabel rules from PD
## Pull down RegionLabel rules from PD

Required for #16141 #16417

`RegionLabel` rules are not yet presently imported from PD. See [TiDB code](https://github.com/pingcap/tidb/blob/3da441d03bb296911b26e33fe9d6712c2044eaa0/br/pkg/pdutil/pd.go#L640) for sample usage.

### PR DESCRIPTION
**Pull request by afeinberg**
### What is changed and how it works?

Issue number: close #16606, ref #16141 

What's Changed:

```commit-message
Pull down RegionLabelRules from PD using MetaStore client.

```

### Related changes

- [ ] PR to update `pingcap/docs`/`pingcap/docs-cn`:
- [ ] Need to cherry-pick to the release branch

### Check List

Tests 

- [X] Unit test
- [ ] Integration test
- [X] Manual test (add detailed scripts or steps below)
Start PD locally, run remove `ignore` and run local_crud_test() in region_label.rs
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: components/region_cache_memory_engine/src/region_label.rs [hunk 1]
```diff
@@ -0,0 +1,406 @@
...
+            pd_client: self.pd_client,
+            path_suffix: self.path_suffix,
+            rule_filter_fn: self.rule_filter_fn,
+        }
+    }
+}
+
+impl RegionLabelService {
+    fn region_label_path(&self) -> String {
+        let cluster_id = self.pd_client.get_cluster_id().unwrap();
```
Comment (nolouch) [inline components/region_cache_memory_engine/src/region_label.rs:132]: the cluster_id will not be changed after cluster startup. so we can initialize it when building the service.

File: components/region_cache_memory_engine/src/region_label.rs [hunk 2]
```diff
@@ -0,0 +1,406 @@
...
+#[derive(Debug, Default, Clone, Serialize, Deserialize, PartialEq)]
+pub struct RegionLabel {
+    pub key: String,
+    pub value: String,
+    pub ttl: Option<String>,
+    pub start_at: Option<String>,
+}
+
+#[derive(Debug, Default, Clone, Serialize, Deserialize, PartialEq)]
+pub struct LabelRule {
```
Comment (nolouch) [inline components/region_cache_memory_engine/src/region_label.rs:27]: I see the structure in Go is 
https://github.com/tikv/pd/blob/783d060861cef37c38cbdcab9777fe95c17907fe/server/schedule/labeler/rules.go#L29C1-L49C1.
is it inconsistent?

Comment (afeinberg) [reply components/region_cache_memory_engine/src/region_label.rs:27]: I'm following this: https://github.com/pingcap/tidb/blob/master/br/pkg/pdutil/pd.go#L648 (note that there's no index specified in pd.go)

Looks like I can remove ttl and start_at from LabelRule. Will do so.

File: components/region_cache_memory_engine/src/region_label.rs (hunk 1 above)
Comment (afeinberg) [reply components/region_cache_memory_engine/src/region_label.rs:132]: done.

File: components/region_cache_memory_engine/src/region_label.rs (hunk 2 above)
Comment (afeinberg) [reply components/region_cache_memory_engine/src/region_label.rs:36]: done.

**Comment by afeinberg** [comment]
Addressed review commments.

File: components/region_cache_memory_engine/src/region_label.rs [hunk 3]
```diff
@@ -0,0 +1,419 @@
...
+    use futures::executor::block_on;
+    use pd_client::meta_storage::{Delete, Put};
+    use security::{SecurityConfig, SecurityManager};
+    use test_pd::{mocker::MetaStorage, util::*, Server as MockServer};
+    use tikv_util::{config::ReadableDuration, worker::Builder};
+
+    use super::*;
+
+    // Note: a test that runs against a local PD instance. This is for debugging
+    // purposesly only and is disabled by default. To run, remove `#[ignore]`
```
Comment (nolouch) [inline components/region_cache_memory_engine/src/region_label.rs:262]:
```suggestion
    // purposes only and is disabled by default. To run, remove `#[ignore]`
```

**Review by nolouch — APPROVED** [review]
lgtm

File: components/region_cache_memory_engine/src/region_label.rs [hunk 4]
```diff
@@ -0,0 +1,419 @@
...
+                            .await;
+                        cancel.abort();
+                        continue 'outer;
+                    }
+                }
+            }
+        }
+    }
+
+    async fn reload_all_region_labels(&mut self) {
```
Comment (SpadeA-Tang) [inline components/region_cache_memory_engine/src/region_label.rs:218]: Should reload clear the current labels first?

Comment (afeinberg) [reply components/region_cache_memory_engine/src/region_label.rs:218]: This will override them. I am matching the behavior with PD used in other cases.

Comment (afeinberg) [reply components/region_cache_memory_engine/src/region_label.rs:218]: Note that this intends to run in a background thread, much like the resource manager does. Will start when server is started.

**Comment by afeinberg** [comment]
Ready for a merge. /cc @tonyxuqqi

**Review by tonyxuqqi — APPROVED** [review]

**Comment by afeinberg** [comment]
@tonyxuqqi the clippy failure looks transient. since it looks like we can't communicate via feishu reliable right now, please do a /merge once we get a clean build.

**Comment by afeinberg** [comment]
Tests passed. Ready for merge @tonyxuqqi

**Pull request merged**